File: config/src/main/java/com/typesafe/config/impl/ConfigImpl.java
/**
 *   Copyright (C) 2011-2012 Typesafe Inc. <http://typesafe.com>
 */
package com.typesafe.config.impl;

import java.io.File;
import java.lang.ref.WeakReference;
import java.net.URL;
import java.time.Duration;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.Properties;
import java.util.concurrent.Callable;

import com.typesafe.config.Config;
import com.typesafe.config.ConfigException;
import com.typesafe.config.ConfigIncluder;
import com.typesafe.config.ConfigMemorySize;
import com.typesafe.config.ConfigObject;
import com.typesafe.config.ConfigOrigin;
import com.typesafe.config.ConfigParseOptions;
import com.typesafe.config.ConfigParseable;
import com.typesafe.config.ConfigValue;
import com.typesafe.config.impl.SimpleIncluder.NameSource;

/**
 * Internal implementation detail, not ABI stable, do not touch.
 * For use only by the {@link com.typesafe.config} package.
 */
public class ConfigImpl {
    private static final String ENV_VAR_OVERRIDE_PREFIX = "CONFIG_FORCE_";

    private static class LoaderCache {
        private Config currentSystemProperties;
        private WeakReference<ClassLoader> currentLoader;
        private Map<String, Config> cache;

        LoaderCache() {
            this.currentSystemProperties = null;
            this.currentLoader = new WeakReference<ClassLoader>(null);
            this.cache = new HashMap<String, Config>();
        }

        // for now, caching as long as the loader remains the same,
        // drop entire cache if it changes.
        synchronized Config getOrElseUpdate(ClassLoader loader, String key, Callable<Config> updater) {
            if (loader != currentLoader.get()) {
                // reset the cache if we start using a different loader
                cache.clear();
                currentLoader = new WeakReference<ClassLoader>(loader);
            }

            Config systemProperties = systemPropertiesAsConfig();
            if (systemProperties != currentSystemProperties) {
                cache.clear();
                currentSystemProperties = systemProperties;
            }

            Config config = cache.get(key);
            if (config == null) {
                try {
                    config = updater.call();
                } catch (RuntimeException e) {
                    throw e; // this will include ConfigException
                } catch (Exception e) {
                    throw new ConfigException.Generic(e.getMessage(), e);
                }
                if (config == null)
                    throw new ConfigException.BugOrBroken("null config from cache updater");
                cache.put(key, config);
            }

            return config;
        }
    }

    private static class LoaderCacheHolder {
        static final LoaderCache cache = new LoaderCache();
    }

    public static Config computeCachedConfig(ClassLoader loader, String key,
            Callable<Config> updater) {
        LoaderCache cache;
        try {
            cache = LoaderCacheHolder.cache;
        } catch (ExceptionInInitializerError e) {
            throw ConfigImplUtil.extractInitializerError(e);
        }
        return cache.getOrElseUpdate(loader, key, updater);
    }


    static class FileNameSource implements SimpleIncluder.NameSource {
        @Override
        public ConfigParseable nameToParseable(String name, ConfigParseOptions parseOptions) {
            return Parseable.newFile(new File(name), parseOptions);
        }
    };

    static class ClasspathNameSource implements SimpleIncluder.NameSource {
        @Override
        public ConfigParseable nameToParseable(String name, ConfigParseOptions parseOptions) {
            return Parseable.newResources(name, parseOptions);
        }
    };

    static class ClasspathNameSourceWithClass implements SimpleIncluder.NameSource {
        final private Class<?> klass;

        public ClasspathNameSourceWithClass(Class<?> klass) {
            this.klass = klass;
        }

        @Override
        public ConfigParseable nameToParseable(String name, ConfigParseOptions parseOptions) {
            return Parseable.newResources(klass, name, parseOptions);
        }
    };

    public static ConfigObject parseResourcesAnySyntax(Class<?> klass, String resourceBasename,
            ConfigParseOptions baseOptions) {
        NameSource source = new ClasspathNameSourceWithClass(klass);
        return SimpleIncluder.fromBasename(source, resourceBasename, baseOptions);
    }

    public static ConfigObject parseResourcesAnySyntax(String resourceBasename,
            ConfigParseOptions baseOptions) {
        NameSource source = new ClasspathNameSource();
        return SimpleIncluder.fromBasename(source, resourceBasename, baseOptions);
    }

    public static ConfigObject parseFileAnySyntax(File basename, ConfigParseOptions baseOptions) {
        NameSource source = new FileNameSource();
        return SimpleIncluder.fromBasename(source, basename.getPath(), baseOptions);
    }

    static AbstractConfigObject emptyObject(String originDescription) {
        ConfigOrigin origin = originDescription != null ? SimpleConfigOrigin
                .newSimple(originDescription) : null;
        return emptyObject(origin);
    }

    public static Config emptyConfig(String originDescription) {
        return emptyObject(originDescription).toConfig();
    }

    static AbstractConfigObject empty(ConfigOrigin origin) {
        return emptyObject(origin);
    }

    // default origin for values created with fromAnyRef and no origin specified
    final private static ConfigOrigin defaultValueOrigin = SimpleConfigOrigin
            .newSimple("hardcoded value");
    final private static ConfigBoolean defaultTrueValue = new ConfigBoolean(
            defaultValueOrigin, true);
    final private static ConfigBoolean defaultFalseValue = new ConfigBoolean(
            defaultValueOrigin, false);
    final private static ConfigNull defaultNullValue = new ConfigNull(
            defaultValueOrigin);
    final private static SimpleConfigList defaultEmptyList = new SimpleConfigList(
            defaultValueOrigin, Collections.<AbstractConfigValue> emptyList());
    final private static SimpleConfigObject defaultEmptyObject = SimpleConfigObject
            .empty(defaultValueOrigin);

    private static SimpleConfigList emptyList(ConfigOrigin origin) {
        if (origin == null || origin == defaultValueOrigin)
            return defaultEmptyList;
        else
            return new SimpleConfigList(origin,
                    Collections.<AbstractConfigValue> emptyList());
    }

    private static AbstractConfigObject emptyObject(ConfigOrigin origin) {
        // we want null origin to go to SimpleConfigObject.empty() to get the
        // origin "empty config" rather than "hardcoded value"
        if (origin == defaultValueOrigin)
            return defaultEmptyObject;
        else
            return SimpleConfigObject.empty(origin);
    }

    private static ConfigOrigin valueOrigin(String originDescription) {
        if (originDescription == null)
            return defaultValueOrigin;
        else
            return SimpleConfigOrigin.newSimple(originDescription);
    }

    public static ConfigValue fromAnyRef(Object object, String originDescription) {
        ConfigOrigin origin = valueOrigin(originDescription);
        return fromAnyRef(object, origin, FromMapMode.KEYS_ARE_KEYS);
    }

    public static ConfigObject fromPathMap(
            Map<String, ? extends Object> pathMap, String originDescription) {
        ConfigOrigin origin = valueOrigin(originDescription);
        return (ConfigObject) fromAnyRef(pathMap, origin,
                FromMapMode.KEYS_ARE_PATHS);
    }

    static AbstractConfigValue fromAnyRef(Object object, ConfigOrigin origin,
            FromMapMode mapMode) {
        if (origin == null)
            throw new ConfigException.BugOrBroken(
                    "origin not supposed to be null");

        if (object == null) {
            if (origin != defaultValueOrigin)
                return new ConfigNull(origin);
            else
                return defaultNullValue;
        } else if(object instanceof AbstractConfigValue) {
            return (AbstractConfigValue) object;
        } else if (object instanceof Boolean) {
            if (origin != defaultValueOrigin) {
                return new ConfigBoolean(origin, (Boolean) object);
            } else if ((Boolean) object) {
                return defaultTrueValue;
            } else {
                return defaultFalseValue;
            }
        } else if (object instanceof String) {
            return new ConfigString.Quoted(origin, (String) object);
        } else if (object instanceof Number) {
            // here we always keep the same type that was passed to us,
            // rather than figuring out if a Long would fit in an Int
            // or a Double has no fractional part. i.e. deliberately
            // not using ConfigNumber.newNumber() when we have a
            // Double, Integer, or Long.
            if (object instanceof Double) {
                return new ConfigDouble(origin, (Double) object, null);
            } else if (object instanceof Integer) {
                return new ConfigInt(origin, (Integer) object, null);
            } else if (object instanceof Long) {
                return new ConfigLong(origin, (Long) object, null);
            } else {
                return ConfigNumber.newNumber(origin,
                        ((Number) object).doubleValue(), null);
            }
        } else if (object instanceof Duration) {
            return new ConfigLong(origin, ((Duration) object).toMillis(), null);
        } else if (object instanceof Map) {
            if (((Map<?, ?>) object).isEmpty())
                return emptyObject(origin);

            if (mapMode == FromMapMode.KEYS_ARE_KEYS) {
                Map<String, AbstractConfigValue> values = new HashMap<String, AbstractConfigValue>();
                for (Map.Entry<?, ?> entry : ((Map<?, ?>) object).entrySet()) {
                    Object key = entry.getKey();
                    if (!(key instanceof String))
                        throw new ConfigException.BugOrBroken(
                                "bug in method caller: not valid to create ConfigObject from map with non-String key: "
                                        + key);
                    AbstractConfigValue value = fromAnyRef(entry.getValue(),
                            origin, mapMode);
                    values.put((String) key, value);
                }

                return new SimpleConfigObject(origin, values);
            } else {
                return PropertiesParser.fromPathMap(origin, (Map<?, ?>) object);
            }
        } else if (object instanceof Iterable) {
            Iterator<?> i = ((Iterable<?>) object).iterator();
            if (!i.hasNext())
                return emptyList(origin);

            List<AbstractConfigValue> values = new ArrayList<AbstractConfigValue>();
            while (i.hasNext()) {
                AbstractConfigValue v = fromAnyRef(i.next(), origin, mapMode);
                values.add(v);
            }

            return new SimpleConfigList(origin, values);
        } else if (object instanceof ConfigMemorySize) {
            return new ConfigLong(origin, ((ConfigMemorySize) object).toBytes(), null);
        } else {
            throw new ConfigException.BugOrBroken(
                    "bug in method caller: not valid to create ConfigValue from: "
                            + object);
        }
    }

    private static class DefaultIncluderHolder {
        static final ConfigIncluder defaultIncluder = new SimpleIncluder(null);
    }

    static ConfigIncluder defaultIncluder() {
        try {
            return DefaultIncluderHolder.defaultIncluder;
        } catch (ExceptionInInitializerError e) {
            throw ConfigImplUtil.extractInitializerError(e);
        }
    }

    private static Properties getSystemProperties() {
        // Avoid ConcurrentModificationException due to parallel setting of system properties by copying properties
        final Properties systemProperties = System.getProperties();
        final Properties systemPropertiesCopy = new Properties();
        synchronized (systemProperties) {
            for (Map.Entry<Object, Object> entry: systemProperties.entrySet()) {
                // Java 11 introduces 'java.version.date', but we don't want that to
                // overwrite 'java.version'
                if (!entry.getKey().toString().startsWith("java.version.")) {
                    systemPropertiesCopy.put(entry.getKey(), entry.getValue());
                }
            }
        }
        return systemPropertiesCopy;
    }

    private static AbstractConfigObject loadSystemProperties() {
        return (AbstractConfigObject) Parseable.newProperties(getSystemProperties(),
                ConfigParseOptions.defaults().setOriginDescription("system properties")).parse();
    }

    private static class SystemPropertiesHolder {
        // this isn't final due to the reloadSystemPropertiesConfig() hack below
        static volatile AbstractConfigObject systemProperties = loadSystemProperties();
    }

    static AbstractConfigObject systemPropertiesAsConfigObject() {
        try {
            return SystemPropertiesHolder.systemProperties;
        } catch (ExceptionInInitializerError e) {
            throw ConfigImplUtil.extractInitializerError(e);
        }
    }

    public static Config systemPropertiesAsConfig() {
        return systemPropertiesAsConfigObject().toConfig();
    }

    public static void reloadSystemPropertiesConfig() {
        // ConfigFactory.invalidateCaches() relies on this having the side
        // effect that it drops all caches
        SystemPropertiesHolder.systemProperties = loadSystemProperties();
    }

    private static AbstractConfigObject loadEnvVariables() {
        return PropertiesParser.fromStringMap(newSimpleOrigin("env variables"), System.getenv());
    }

    private static class EnvVariablesHolder {
        static volatile AbstractConfigObject envVariables = loadEnvVariables();
    }

    static AbstractConfigObject envVariablesAsConfigObject() {
        try {
            return EnvVariablesHolder.envVariables;
        } catch (ExceptionInInitializerError e) {
            throw ConfigImplUtil.extractInitializerError(e);
        }
    }

    public static Config envVariablesAsConfig() {
        return envVariablesAsConfigObject().toConfig();
    }

    public static void reloadEnvVariablesConfig() {
        // ConfigFactory.invalidateCaches() relies on this having the side
        // effect that it drops all caches
        EnvVariablesHolder.envVariables = loadEnvVariables();
    }



    private static AbstractConfigObject loadEnvVariablesOverrides() {
        Map<String, String> env = new HashMap(System.getenv());
        Map<String, String> result = new HashMap(System.getenv());

        for (String key : env.keySet()) {
            if (key.startsWith(ENV_VAR_OVERRIDE_PREFIX)) {
                result.put(ConfigImplUtil.envVariableAsProperty(key, ENV_VAR_OVERRIDE_PREFIX), env.get(key));
            }
        }

        return PropertiesParser.fromStringMap(newSimpleOrigin("env variables overrides"), result);
    }

    private static class EnvVariablesOverridesHolder {
        static volatile AbstractConfigObject envVariables = loadEnvVariablesOverrides();
    }

    static AbstractConfigObject envVariablesOverridesAsConfigObject() {
        try {
            return EnvVariablesOverridesHolder.envVariables;
        } catch (ExceptionInInitializerError e) {
            throw ConfigImplUtil.extractInitializerError(e);
        }
    }

    public static Config envVariablesOverridesAsConfig() {
        return envVariablesOverridesAsConfigObject().toConfig();
    }

    public static void reloadEnvVariablesOverridesConfig() {
        // ConfigFactory.invalidateCaches() relies on this having the side
        // effect that it drops all caches
        EnvVariablesOverridesHolder.envVariables = loadEnvVariablesOverrides();
    }

    public static Config defaultReference(final ClassLoader loader) {
        return computeCachedConfig(loader, "defaultReference", new Callable<Config>() {
            @Override
            public Config call() {
                Config unresolvedResources = unresolvedReference(loader);
                return systemPropertiesAsConfig().withFallback(unresolvedResources).resolve();
            }
        });
    }

    private static Config unresolvedReference(final ClassLoader loader) {
        return computeCachedConfig(loader, "unresolvedReference", new Callable<Config>() {
            @Override
            public Config call() {
                return Parseable.newResources("reference.conf",
                        ConfigParseOptions.defaults().setClassLoader(loader))
                    .parse().toConfig();
            }
        });
    }

    /**
     * This returns the unresolved reference configuration, but before doing so,
     * it verifies that the reference configuration resolves, to ensure that it
     * is self contained and doesn't depend on any higher level configuration
     * files.
     */
    public static Config defaultReferenceUnresolved(final ClassLoader loader) {
        // First, verify that `reference.conf` resolves by itself.
        try {
            defaultReference(loader);
        } catch (ConfigException.UnresolvedSubstitution e) {
            throw e.addExtraDetail("Could not resolve substitution in reference.conf to a value: %s. All reference.conf files are required to be fully, independently resolvable, and should not require the presence of values for substitutions from further up the hierarchy.");
        }
        // Now load the unresolved version
        return unresolvedReference(loader);
    }


    private static class DebugHolder {
        private static String LOADS = "loads";
        private static String SUBSTITUTIONS = "substitutions";

        private static Map<String, Boolean> loadDiagnostics() {
            Map<String, Boolean> result = new HashMap<String, Boolean>();
            result.put(LOADS, false);
            result.put(SUBSTITUTIONS, false);

            // People do -Dconfig.trace=foo,bar to enable tracing of different things
            String s = System.getProperty("config.trace");
            if (s == null) {
                return result;
            } else {
                String[] keys = s.split(",");
                for (String k : keys) {
                    if (k.equals(LOADS)) {
                        result.put(LOADS, true);
                    } else if (k.equals(SUBSTITUTIONS)) {
                        result.put(SUBSTITUTIONS, true);
                    } else {
                        System.err.println("config.trace property contains unknown trace topic '"
                                + k + "'");
                    }
                }
                return result;
            }
        }

        private static final Map<String, Boolean> diagnostics = loadDiagnostics();

        private static final boolean traceLoadsEnabled = diagnostics.get(LOADS);
        private static final boolean traceSubstitutionsEnabled = diagnostics.get(SUBSTITUTIONS);

        static boolean traceLoadsEnabled() {
            return traceLoadsEnabled;
        }

        static boolean traceSubstitutionsEnabled() {
            return traceSubstitutionsEnabled;
        }
    }

    public static boolean traceLoadsEnabled() {
        try {
            return DebugHolder.traceLoadsEnabled();
        } catch (ExceptionInInitializerError e) {
            throw ConfigImplUtil.extractInitializerError(e);
        }
    }

    public static boolean traceSubstitutionsEnabled() {
        try {
            return DebugHolder.traceSubstitutionsEnabled();
        } catch (ExceptionInInitializerError e) {
            throw ConfigImplUtil.extractInitializerError(e);
        }
    }

    public static void trace(String message) {
        System.err.println(message);
    }

    public static void trace(int indentLevel, String message) {
        while (indentLevel > 0) {
            System.err.print("  ");
            indentLevel -= 1;
        }
        System.err.println(message);
    }

    // the basic idea here is to add the "what" and have a canonical
    // toplevel error message. the "original" exception may however have extra
    // detail about what happened. call this if you have a better "what" than
    // further down on the stack.
    static ConfigException.NotResolved improveNotResolved(Path what,
            ConfigException.NotResolved original) {
        String newMessage = what.render()
                + " has not been resolved, you need to call Config#resolve(),"
                + " see API docs for Config#resolve()";
        if (newMessage.equals(original.getMessage()))
            return original;
        else
            return new ConfigException.NotResolved(newMessage, original);
    }

    public static ConfigOrigin newSimpleOrigin(String description) {
        if (description == null) {
            return defaultValueOrigin;
        } else {
            return SimpleConfigOrigin.newSimple(description);
        }
    }

    public static ConfigOrigin newFileOrigin(String filename) {
        return SimpleConfigOrigin.newFile(filename);
    }

    public static ConfigOrigin newURLOrigin(URL url) {
        return SimpleConfigOrigin.newURL(url);
    }
}

File: config/src/main/scala/com/typesafe/config/impl/ConfigImpl.scala
/**
 *   Copyright (C) 2011-2012 Typesafe Inc. <http://typesafe.com>
 */
package com.typesafe.config.impl

import java.io.File
import java.{ lang => jl }
import jl.ref.WeakReference
import java.net.URL
import java.time.Duration
import java.{ util => ju }
import java.util.concurrent.Callable
import scala.collection.JavaConverters._
import com.typesafe.config.Config
import com.typesafe.config.ConfigException
import com.typesafe.config.ConfigIncluder
import com.typesafe.config.ConfigMemorySize
import com.typesafe.config.ConfigObject
import com.typesafe.config.ConfigOrigin
import com.typesafe.config.ConfigParseOptions
import com.typesafe.config.ConfigParseable
import com.typesafe.config.ConfigValue

/**
 * Internal implementation detail, not ABI stable, do not touch.
 * For use only by the {@link com.typesafe.config} package.
 */
object ConfigImpl {

    private class LoaderCache private[impl] () {
        private[impl] var currentSystemProperties: Config = null
        private var currentLoader = new WeakReference[ClassLoader](null)
        private var cache = new ju.HashMap[String, Config]

        // for now, caching as long as the loader remains the same,
        // drop entire cache if it changes.
        private[impl] def getOrElseUpdate(
            loader: ClassLoader,
            key: String,
            updater: Callable[Config]): Config = synchronized {
            if (loader != currentLoader.get) {
                // reset the cache if we start using a different loader
                cache.clear()
                currentLoader = new WeakReference[ClassLoader](loader)
            }
            val systemProperties = systemPropertiesAsConfig
            if (systemProperties != currentSystemProperties) {
                cache.clear()
                currentSystemProperties = systemProperties
            }
            var config = cache.get(key)
            if (config == null) {
                try config = updater.call
                catch {
                    case e: RuntimeException =>
                        throw e // this will include ConfigException

                    case e: Exception =>
                        throw new ConfigException.Generic(e.getMessage, e)
                }
                if (config == null) throw new ConfigException.BugOrBroken(
                    "null config from cache updater")
                cache.put(key, config)
            }
            config
        }
    }
    private object LoaderCacheHolder {
        private[impl] val cache = new ConfigImpl.LoaderCache
        private[impl] def invalidate: Unit = cache.synchronized {
            cache.currentSystemProperties = null
        }
    }

    def computeCachedConfig(
        loader: ClassLoader,
        key: String,
        updater: Callable[Config]): Config = {
        var cache: LoaderCache = null
        try cache = LoaderCacheHolder.cache
        catch {
            case e: ExceptionInInitializerError =>
                throw ConfigImplUtil.extractInitializerError(e)
        }
        cache.getOrElseUpdate(loader, key, updater)
    }
    private[impl] class FileNameSource extends SimpleIncluder.NameSource {
        override def nameToParseable(
            name: String,
            parseOptions: ConfigParseOptions): ConfigParseable = Parseable.newFile(new File(name), parseOptions)
    }
    private[impl] class ClasspathNameSource extends SimpleIncluder.NameSource {
        override def nameToParseable(
            name: String,
            parseOptions: ConfigParseOptions): ConfigParseable = Parseable.newResources(name, parseOptions)
    }
    private[impl] class ClasspathNameSourceWithClass(val klass: Class[_])
        extends SimpleIncluder.NameSource {
        override def nameToParseable(
            name: String,
            parseOptions: ConfigParseOptions): ConfigParseable = Parseable.newResources(klass, name, parseOptions)
    }
    def parseResourcesAnySyntax(
        klass: Class[_],
        resourceBasename: String,
        baseOptions: ConfigParseOptions): ConfigObject = {
        val source =
            new ConfigImpl.ClasspathNameSourceWithClass(klass)
        SimpleIncluder.fromBasename(source, resourceBasename, baseOptions)
    }
    def parseResourcesAnySyntax(
        resourceBasename: String,
        baseOptions: ConfigParseOptions): ConfigObject = {
        val source = new ConfigImpl.ClasspathNameSource
        SimpleIncluder.fromBasename(source, resourceBasename, baseOptions)
    }
    def parseFileAnySyntax(
        basename: File,
        baseOptions: ConfigParseOptions): ConfigObject = {
        val source = new ConfigImpl.FileNameSource
        SimpleIncluder.fromBasename(source, basename.getPath, baseOptions)
    }
    private[impl] def emptyObject(
        originDescription: String): AbstractConfigObject = {
        val origin =
            if (originDescription != null) SimpleConfigOrigin.newSimple(originDescription) else null
        emptyObject(origin)
    }
    def emptyConfig(originDescription: String): Config = emptyObject(originDescription).toConfig

    private[impl] def empty(origin: ConfigOrigin): AbstractConfigObject = emptyObject(origin)
    // default origin for values created with fromAnyRef and no origin specified
    private val defaultValueOrigin =
        SimpleConfigOrigin.newSimple("hardcoded value")
    private val defaultTrueValue =
        new ConfigBoolean(defaultValueOrigin, true)
    private val defaultFalseValue =
        new ConfigBoolean(defaultValueOrigin, false)
    private val defaultNullValue = new ConfigNull(defaultValueOrigin)
    private val defaultEmptyList = new SimpleConfigList(
        defaultValueOrigin,
        ju.Collections.emptyList[AbstractConfigValue])
    private val defaultEmptyObject =
        SimpleConfigObject.empty(defaultValueOrigin)
    private def emptyList(origin: ConfigOrigin): SimpleConfigList =
        if (origin == null || (origin == defaultValueOrigin)) defaultEmptyList
        else new SimpleConfigList(
            origin,
            ju.Collections.emptyList[AbstractConfigValue])
    private def emptyObject(origin: ConfigOrigin): AbstractConfigObject = {
        // we want null origin to go to SimpleConfigObject.empty() to get the
        // origin "empty config" rather than "hardcoded value"
        if (origin == defaultValueOrigin) defaultEmptyObject else SimpleConfigObject.empty(origin)
    }
    private def valueOrigin(originDescription: String): ConfigOrigin =
        if (originDescription == null) defaultValueOrigin
        else SimpleConfigOrigin.newSimple(originDescription)
    def fromAnyRef(`object`: AnyRef, originDescription: String): ConfigValue = {
        val origin = valueOrigin(originDescription)
        fromAnyRef(`object`, origin, FromMapMode.KEYS_ARE_KEYS)
    }
    def fromPathMap(
        pathMap: ju.Map[String, _],
        originDescription: String): ConfigObject = {
        val origin = valueOrigin(originDescription)
        fromAnyRef(pathMap, origin, FromMapMode.KEYS_ARE_PATHS)
            .asInstanceOf[ConfigObject]
    }
    def fromAnyRef(`object`: Any, origin: ConfigOrigin, mapMode: FromMapMode): AbstractConfigValue = {
        if (origin == null)
            throw new ConfigException.BugOrBroken("origin not supposed to be null")
        if (`object` == null)
            if (origin != defaultValueOrigin)
                new ConfigNull(origin)
            else defaultNullValue
        else if (`object`.isInstanceOf[AbstractConfigValue])
            `object`.asInstanceOf[AbstractConfigValue]
        else if (`object`.isInstanceOf[jl.Boolean])
            if (origin != defaultValueOrigin)
                new ConfigBoolean(origin, `object`.asInstanceOf[jl.Boolean])
            else if (`object`.asInstanceOf[jl.Boolean]) defaultTrueValue
            else defaultFalseValue
        else if (`object`.isInstanceOf[String])
            new ConfigString.Quoted(origin, `object`.asInstanceOf[String])
        else if (`object`.isInstanceOf[Number]) {
            // here we always keep the same type that was passed to us,
            // rather than figuring out if a Long would fit in an Int
            // or a Double has no fractional part. i.e. deliberately
            // not using ConfigNumber.newNumber() when we have a
            // Double, Integer, or Long.
            if (`object`.isInstanceOf[jl.Double])
                new ConfigDouble(origin, `object`.asInstanceOf[jl.Double], null)
            else if (`object`.isInstanceOf[Integer])
                new ConfigInt(origin, `object`.asInstanceOf[Integer], null)
            else if (`object`.isInstanceOf[jl.Long])
                new ConfigLong(origin, `object`.asInstanceOf[jl.Long], null)
            else
                ConfigNumber.newNumber(
                    origin,
                    `object`.asInstanceOf[Number].doubleValue,
                    null)
        } else if (`object`.isInstanceOf[Duration]) {
            new ConfigLong(origin, `object`.asInstanceOf[Duration].toMillis, null)
        } else if (`object`.isInstanceOf[ju.Map[_, _]]) {
            if (`object`.asInstanceOf[ju.Map[_, _]].isEmpty)
                return emptyObject(origin)
            if (mapMode == FromMapMode.KEYS_ARE_KEYS) {
                val values = new ju.HashMap[String, AbstractConfigValue]
                for (entry <- `object`.asInstanceOf[ju.Map[_, _]].entrySet.asScala) {
                    val key = entry.getKey
                    if (!key.isInstanceOf[String])
                        throw new ConfigException.BugOrBroken(
                            "bug in method caller: not valid to create ConfigObject from map with non-String key: " + key)
                    val value = fromAnyRef(entry.getValue, origin, mapMode)
                    values.put(key.asInstanceOf[String], value)
                }
                new SimpleConfigObject(origin, values)
            } else {
                PropertiesParser.fromPathMap(origin, `object`.asInstanceOf[ju.Map[_, _]])
            }
        } else if (`object`.isInstanceOf[jl.Iterable[_]]) {
            val i = `object`.asInstanceOf[jl.Iterable[_]].iterator
            if (!i.hasNext) return emptyList(origin)
            val values = new ju.ArrayList[AbstractConfigValue]
            while (i.hasNext) {
                val v = fromAnyRef(i.next, origin, mapMode)
                values.add(v)
            }
            new SimpleConfigList(origin, values)
        } else if (`object`.isInstanceOf[ConfigMemorySize]) {
            new ConfigLong(
                origin,
                `object`.asInstanceOf[ConfigMemorySize].toBytes,
                null)
        } else {
            throw new ConfigException.BugOrBroken(
                "bug in method caller: not valid to create ConfigValue from: " + `object`)
        }
    }
    private object DefaultIncluderHolder {
        private[impl] val defaultIncluder = new SimpleIncluder(null)
    }
    private[impl] def defaultIncluder: ConfigIncluder =
        try DefaultIncluderHolder.defaultIncluder
        catch {
            case e: ExceptionInInitializerError =>
                throw ConfigImplUtil.extractInitializerError(e)
        }
    private def getSystemProperties: ju.Properties = {
        // Avoid ConcurrentModificationException due to parallel setting of system properties by copying properties
        val systemProperties = System.getProperties
        val systemPropertiesCopy = new ju.Properties
        systemProperties.synchronized {
            systemPropertiesCopy.putAll(systemProperties)
        }
        systemPropertiesCopy
    }
    private def loadSystemProperties: AbstractConfigObject =
        Parseable
            .newProperties(
                getSystemProperties,
                ConfigParseOptions.defaults.setOriginDescription("system properties"))
            .parse
            .asInstanceOf[AbstractConfigObject]
    private object SystemPropertiesHolder {
        // this isn't final due to the reloadSystemPropertiesConfig() hack below
        @volatile private[impl] var systemProperties: AbstractConfigObject = loadSystemProperties
    }
    private[impl] def systemPropertiesAsConfigObject: AbstractConfigObject =
        try SystemPropertiesHolder.systemProperties
        catch {
            case e: ExceptionInInitializerError =>
                throw ConfigImplUtil.extractInitializerError(e)
        }
    def systemPropertiesAsConfig: Config = systemPropertiesAsConfigObject.toConfig
    def reloadSystemPropertiesConfig(): Unit = {
        // ConfigFactory.invalidateCaches() relies on this having the side
        // effect that it drops all caches
        // change - could not find the side effect so added a method to invalidate explicitly
        LoaderCacheHolder.invalidate
        SystemPropertiesHolder.systemProperties = loadSystemProperties
    }
    private def loadEnvVariables: AbstractConfigObject = PropertiesParser.fromStringMap(
        newSimpleOrigin("env variables"),
        System.getenv)
    private object EnvVariablesHolder {
        @volatile private[impl] var envVariables = loadEnvVariables
    }
    def envVariablesAsConfigObject: AbstractConfigObject = try EnvVariablesHolder.envVariables
    catch {
        case e: ExceptionInInitializerError =>
            throw ConfigImplUtil.extractInitializerError(e)
    }
    def envVariablesAsConfig: Config = envVariablesAsConfigObject.toConfig
    def reloadEnvVariablesConfig(): Unit = {
        // ConfigFactory.invalidateCaches() relies on this having the side
        // effect that it drops all caches
        EnvVariablesHolder.envVariables = loadEnvVariables
    }
    def defaultReference(loader: ClassLoader): Config = computeCachedConfig(
        loader,
        "defaultReference",
        new Callable[Config]() {
            override def call: Config = {
                val unresolvedResources = Parseable
                    .newResources(
                        "reference.conf",
                        ConfigParseOptions.defaults.setClassLoader(loader))
                    .parse
                    .toConfig
                systemPropertiesAsConfig
                    .withFallback(unresolvedResources)
                    .resolve
            }
        })
    private object DebugHolder {
        private val LOADS = "loads"
        private val SUBSTITUTIONS = "substitutions"
        private def loadDiagnostics: ju.Map[String, jl.Boolean] = {
            val result = new ju.HashMap[String, jl.Boolean]
            result.put(LOADS, false)
            result.put(SUBSTITUTIONS, false)
            // People do -Dconfig.trace=foo,bar to enable tracing of different things
            val s = System.getProperty("config.trace")
            if (s == null) result
            else {
                val keys = s.split(",")
                for (k <- keys) {
                    if (k == LOADS)
                        result.put(LOADS, true)
                    else if (k == SUBSTITUTIONS)
                        result.put(SUBSTITUTIONS, true)
                    else
                        System.err.println(
                            "config.trace property contains unknown trace topic '" + k + "'")
                }
                result
            }
        }
        private[impl] val diagnostics: ju.Map[String, jl.Boolean] = loadDiagnostics
        private[impl] val traceLoadsEnabled: Boolean = diagnostics.get(LOADS)
        private[impl] val traceSubstitutionsEnabled: Boolean = diagnostics.get(SUBSTITUTIONS)
        // two methods for the two previous vals are not needed in Scala
    }

    def traceLoadsEnabled: Boolean =
        try DebugHolder.traceLoadsEnabled
        catch {
            case e: ExceptionInInitializerError =>
                throw ConfigImplUtil.extractInitializerError(e)
        }
    def traceSubstitutionsEnabled: Boolean =
        try DebugHolder.traceSubstitutionsEnabled
        catch {
            case e: ExceptionInInitializerError =>
                throw ConfigImplUtil.extractInitializerError(e)
        }
    def trace(message: String): Unit = System.err.println(message)
    def trace(indentLevel: Int, message: String): Unit = {
        var level = indentLevel
        while (level > 0) {
            System.err.print("  ")
            level -= 1
        }
        System.err.println(message)
    }
    // the basic idea here is to add the "what" and have a canonical
    // toplevel error message. the "original" exception may however have extra
    // detail about what happened. call this if you have a better "what" than
    // further down on the stack.
    def improveNotResolved(
        what: Path,
        original: ConfigException.NotResolved): ConfigException.NotResolved = {
        val newMessage = what.render +
            " has not been resolved, you need to call Config#resolve()," +
            " see API docs for Config#resolve()"
        if (newMessage == original.getMessage) original
        else new ConfigException.NotResolved(newMessage, original)
    }
    def newSimpleOrigin(description: String): ConfigOrigin =
        if (description == null) defaultValueOrigin
        else SimpleConfigOrigin.newSimple(description)
    def newFileOrigin(filename: String): ConfigOrigin = SimpleConfigOrigin.newFile(filename)
    def newURLOrigin(url: URL): ConfigOrigin = SimpleConfigOrigin.newURL(url)
}
